File: src/lib/backend-client/remote.ts
import type {
  RemoteBackupPreview,
  RemoteBackupResult,
  RemoteBackupVerification,
  S3CredentialInput,
} from '../types'
import { call } from './shared'

export const remoteClient = {
  storeCredentials: (credentials: S3CredentialInput) =>
    call<void>('store_s3_credentials', { credentials }),
  clearCredentials: () => call<void>('clear_s3_credentials'),
  previewBackup: () => call<RemoteBackupPreview>('preview_remote_backup'),
  runBackup: () => call<RemoteBackupResult>('run_remote_backup'),
  verifyBackup: (bundlePath: string) =>
    call<RemoteBackupVerification>('verify_remote_backup', { bundlePath }),
}

File: src-tauri/src/worker_bridge/remote.rs
use vault_core::S3CredentialInput;

use super::worker_result;

pub(crate) fn preview_remote_backup_impl() -> Result<vault_core::RemoteBackupPreview, String> {
    worker_result(vault_worker::preview_remote_backup_bundle())
}

pub(crate) fn run_remote_backup_impl(
    session_database_key: Option<&str>,
) -> Result<vault_core::RemoteBackupResult, String> {
    worker_result(vault_worker::upload_remote_backup_bundle(session_database_key))
}

#[cfg_attr(test, allow(dead_code))]
pub(crate) fn verify_remote_backup_impl(
    bundle_path: String,
    session_database_key: Option<&str>,
) -> Result<vault_core::RemoteBackupVerification, String> {
    worker_result(vault_worker::verify_remote_backup_bundle(session_database_key, &bundle_path))
}

pub(crate) fn store_s3_credentials_impl(credentials: S3CredentialInput) -> Result<(), String> {
    worker_result(vault_worker::store_s3_credentials(&credentials))
}

pub(crate) fn clear_s3_credentials_impl() -> Result<(), String> {
    worker_result(vault_worker::clear_s3_credentials())
}
